File: leetcode/removeNthFromEnd.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    let behind = head;
    let ahead = head;
    for (let i = 0; i < n; i++) {
        ahead = ahead.next;
    }
    if (ahead === null) return head.next;
    ahead = ahead.next;
    while (ahead !== null) {
        behind = behind.next;
        ahead = ahead.next;
    }
    behind.next = behind.next.next;
    return head;
};

File: leetcode/removeNthFromEnd.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        ListNode curr = dummy;
        ListNode ahead = dummy;
        curr.next = head;
        while (n-- >= 0) {
            ahead = ahead.next;
        }
        while (ahead != null) {
            curr = curr.next;
            ahead = ahead.next;
        }
        curr.next = curr.next.next;
        return dummy.next;
    }
}